### ISSUE TITLE
Project goal(s) / Contributions

### ISSUE THREAD

**Issue by nyhu** [EmeraldCityCircusArts/adp#1]
Hello @mikecook 

I have a need for an simple adp client and time to contribute.

I wanna collect infos (workers and salary) from adp. I have some knowledge around Oauth2 but I miss two things.
- A dev env, adp takes a lot of time to answer and the api is still close for me at this time, you might have an shareable access to their staging env ?
- A project to share my effort with. Today you are the only one on the market in golang, it's still better than python alternative which is odd. But it seems you did not had time to grow the project beside data models.

Are you still in need of this client today ? Do you have time to review / plan evolution of the project ?

Regards from France

**Comment by mikecook** [comment]
Hi @nyhu, thank you for your interest!

I have a client, and it's in use running in AWS Lambda, but I was time constrained and included some things I need to refactor out into separate repos, and some config that should be separated from the code before I share it. I've created a quick and dirty client gist for you that I haven't tested for completeness, but it is extracted from working code:  https://gist.github.com/mikecook/07199a2d20f6087f7daa626bc382b911

There are 4 credentials. 
- cert (leaf, intermediate, root)
- cert private key
- client ID
- client Secret

You generate the CSR and cert private key and they sign the CSR and return you a leaf, intermediate, and root cert using this howto: https://developers.adp.com/articles/general/generate-a-certificate-signing-request When you sign a support contract with ADP that specifically includes API access they will issue you a Client ID and a Client Secret.

I am unable to share my access with you, though there may be things I can test if I have time. I should also clarify that this is a bit of a side project for me, I have a separate full time job that takes the vast majority of my attention. When I do find some time, after adding a proper client, and pushing my calendaring work, I see the new golang generics benefiting the model helper methods. I'd rather an OpenAPI schema to generate go bindings from, but ADP doesn't provide that and creating the schema myself from the json schema, meta schema, and postman collections was going to be more time consuming than just generating the golang models from the json schema.

Right now the code I have hits the worker, workSchedule, and workerManagement APIs to find out all the worker schedules for this organization and pushes those schedules into departmental google calendars, purely to simplify scheduling for this org. I'm not using the ADP meta APIs for validation like I should, though I wrote some code to use them for troubleshooting. I do not have access to the salary API because of organizational policy.

I should note that I know of at least one company that does ADP API stuff and provides a golang API, or at least OpenAPI schemas: https://merge.dev , who make a meta API that provides a standardized API across multiple services. Full disclosure, I interviewed with them after I had created this but they ultimately chose another candidate. They are passionate and very cool. I recommend. Here's one of their go apis on github https://github.com/merge-api/merge-hris-go

If there's collaboration and interest I'll find more time to make this project better.

**Comment by nyhu** [comment]
Hello @mikecook,

Thanks a lot for your answer, `merge.dev` is a good recommendation, I scheduled a demo with them next week. I really enjoy their HR portfolio that cover a large part of our integration needs.

As we are not directly client of adp, it would be a pain for us to test / add value to this project. So we interviewed with a company that have privileged access to adp, and can even bypass the api if needed (founded by a former adp employee). Their price is suitable for our need and we will do a poc before going further in our collaboration. 

I think I can now see the how and why for this project, we have kinda different need about adp api, different but complementary. I don't think that in my position I can contribute today to this project since I can't setup a correct dev environment. I'm not sure that your need would grow in the nearest future. And I don't want to ask for features you don't require yourself! 

I would gladly be in touch we you, mostly about interacting with companies calendars which is another, even more generic need that we share. We could meet sometime this week, feel free to contact me by [email](https://github.com/nyhu).

Cheers.
